### ISSUE TITLE
[FEATURE] Implement NIP-26 delegated key login via QR codes

### ISSUE THREAD

**Issue by sethforprivacy** [vitorpamplona/amethyst#331]
**Is your feature request related to a problem? Please describe.**
Currently when using Nostr, you have to expose your root private key to any client you wish to use with Nostr, including Amethyst. This opens up attack vectors that would expose your entire Nostr identity to being stolen or compromised.

A solution has been proposed in NIP-26, allowing you to delegate signing to new keypairs (likely one set of keys per client for granular control) by signing a delegation tag with the root key.

Leveraging this proposal allows you to still use Amethyst as normal but prevent your root Nostr key from ever being exposed.

**Describe the solution you'd like**
Ideally, we will allow the delegation event signing flow to be handled as an alternative log-in method, with the ideal flow for me looking like this:

1. User opens Amethyst and is prompted with the login screen
2. An option appears above "Generate a new key" that says something about "Login with delegated keys"
3. A new screen appears, allowing a user to select the duration of the key delegation and the allowed permissions
4. Amethyst prompts the user to scan an npub via QR code
5. The user's hardware wallet displays their npub as a QR code, and Amethyst scans it
6. Amethyst populates a QR code with the delegation string and displays it for scanning by another device, like a hardware wallet
7. The hardware wallet processes the delegation string and displays the details on-screen
8. If the user approves of the delegation string, the hardware wallet creates a SHA256 hash of the string, signs it, and creates a QR code with the delegation token
9. Amethyst opens a QR code scanner, allowing the user to scan in the delegation token
10. After Amethyst validates the delegation token, it publishes the delegation note of kind 1 to a default set of relays and logs in as normal

As this allows granular permissions, we could implement a simple MVP that allows full access to delegatee's and implement granular permissions later with the necessary UI/UX changes to account for Amethyst not having permissions to perform certain actions.

**Bounty (in Bitcoin sats) offered for the implementation**
I am offering a **1.5m sat** bounty for this feature to be implemented in Amethyst.

**Total bounty (as of April 4th): 3.5m sats**

**Additional context**
We will be working on an implementation of this flow for Passport, our hardware wallet, and I will provide a demo of the proposed flow as soon as we have that up and working.

For details on NIP-26 generally, you can see the specification here:

https://github.com/nostr-protocol/nips/blob/master/26.md

**Comment by sethforprivacy** [comment]
As proper testing of this bounty during implementation really requires a compatible hardware wallet, we'd love to offer a free [Passport](https://foundationdevices.com/passport/) to the dev that takes this on, dependent on some proof that they have the ability to complete the task required (i.e. existing commits to Amethyst or similar Nostr apps).

Perhaps @vitorpamplona or @greenart7c3 would want to tackle this, or know of someone who would be a good candidate? Both of you obviously have shown your ability to implement this :smile:

**Comment by sethforprivacy** [comment]
An additional 2m sats have been pledged by an anonymous Nostrich, which I will pay out upon completion from my own wallet.

**Comment by vitorpamplona** [comment]
This is slightly larger than usual. If somebody wants to take this, let people know here that you are working on it. I am sure, @sethforprivacy can split the bounty if people are working together on a single solution.

**Comment by sethforprivacy** [comment]
For sure, if contributors are willing to come up with a split I'm happy to payout accordingly.

**Comment by sethforprivacy** [comment]
Note that a similar workflow has been explored by @kdmukai for SeedSigner here:

https://gist.github.com/kdmukai/089ff308506058973974479ae4b64b42

**Comment by mjg-foundation** [comment]
I'll be working on the Passport side of this, so if anyone working on this wants to discuss the integration, I'd be happy to!

**Comment by alltheseas** [comment]
> Note that a similar workflow has been explored by @kdmukai for SeedSigner here:
> 
> 
> 
> https://gist.github.com/kdmukai/089ff308506058973974479ae4b64b42

Yes, and building on top of @kdmukai's seedsigner implementation here is the user story drafted a few weeks back: https://github.com/damus-io/damus/issues/451.

**Comment by greenart7c3** [comment]
I'll take this. Talked to @sethforprivacy and when the [Passport](https://foundationdevices.com/passport/) device arrives i'll start working on it.

**Comment by mjg-foundation** [comment]
> I'll take this. Talked to @sethforprivacy and when the [Passport](https://foundationdevices.com/passport/) device arrives i'll start working on it.

Here's the current state of the UI. This isn't necessarily final, and is open to feedback. I tried to make it similar enough to @kdmukai 's UI flow, so any clients supporting this would work with both.

![snapshot-$j-12:51:10](https://user-images.githubusercontent.com/114431859/234077331-1230483e-43a4-46fb-a601-24a854ce78e1.png)
![snapshot-$j-12:51:13](https://user-images.githubusercontent.com/114431859/234077333-ae2adfc8-bfca-42ec-bd4e-d7f2842f9299.png)
![snapshot-$j-12:51:17](https://user-images.githubusercontent.com/114431859/234077337-cabf437c-3422-4fee-a530-892011a55060.png)
![snapshot-$j-12:51:20](https://user-images.githubusercontent.com/114431859/234077339-9dbade06-adc7-44ca-bc08-3890efe76cd0.png)
![snapshot-$j-12:51:25](https://user-images.githubusercontent.com/114431859/234077341-06ac7e34-f4c7-4f5c-873e-9960352f9a93.png)
![snapshot-$j-12:51:28](https://user-images.githubusercontent.com/114431859/234077343-76e435c9-0b6e-41d8-95ac-7a366b561796.png)
![snapshot-$j-12:51:30](https://user-images.githubusercontent.com/114431859/234077346-ea7ae5d4-58eb-41cc-a5e8-6a1d14027d3d.png)
![snapshot-$j-12:51:34](https://user-images.githubusercontent.com/114431859/234077347-afe89d0b-24bd-4b78-bfda-c1e996af15e1.png)
![snapshot-$j-12:52:44](https://user-images.githubusercontent.com/114431859/234077349-42de86ef-4d8d-4342-9af8-33808fa062a8.png)
![snapshot-$j-12:53:23](https://user-images.githubusercontent.com/114431859/234077350-fc9db09e-0722-41c3-bd25-0425956479de.png)
![snapshot-$j-12:53:40](https://user-images.githubusercontent.com/114431859/234077351-84722fd0-ecd7-4e57-b962-3e7bdae4d773.png)
![snapshot-$j-12:53:46](https://user-images.githubusercontent.com/114431859/234077352-6e64f263-3523-4fd2-b093-832d0d95e6ff.png)
![snapshot-$j-12:53:49](https://user-images.githubusercontent.com/114431859/234077354-6f5bf398-1564-4163-937f-651c0afd8011.png)

I also made and enforced some assumptions about the format of the delegation string:
- There will be at most one "created_at<", and at most one "created_at>", in each delegation string, which apply to all kinds included in the string.
- All "kind"s must come before all "created_at"s
- The "created_at>" timestamp must be less than the "created_at<" timestamp

I think the NIP26 spec is a bit unclear as to whether multiple sets of kinds with different created_at times can be delegated in one go, so I assumed not. If this is not the case, I'd be happy to change my assumptions.

**Comment by mjg-foundation** [comment]
Here are the pages specific to using the microSD, skipping the pages shared by both methods. Seth didn't mention it in the initial bounty, so I'll pledge another 100k sats for interoperability with this.

![snapshot-$j-13:10:46](https://user-images.githubusercontent.com/114431859/234080817-9dbb2083-b3de-4631-b19d-5bd42cb42fa0.png)
![snapshot-$j-13:10:51](https://user-images.githubusercontent.com/114431859/234080820-3397a30b-181f-4614-ad19-b5595faaa98e.png)
![snapshot-$j-13:10:57](https://user-images.githubusercontent.com/114431859/234080821-50e6e97a-b9fb-4080-ab94-55f60cff9288.png)
![snapshot-$j-13:11:00](https://user-images.githubusercontent.com/114431859/234080823-3d1ee27e-217c-4f40-8ffe-081333b17a61.png)
![snapshot-$j-13:11:05](https://user-images.githubusercontent.com/114431859/234080826-7fb00a41-0c7f-4120-9ff8-a485df734ceb.png)
![snapshot-$j-13:11:18](https://user-images.githubusercontent.com/114431859/234080827-65fc9766-c14f-4b77-a6fc-9220be3b530d.png)
![snapshot-$j-13:11:22](https://user-images.githubusercontent.com/114431859/234080828-5161a349-bc51-47f0-a237-1ddb8d62bf78.png)
![snapshot-$j-13:11:28](https://user-images.githubusercontent.com/114431859/234080829-d39d322c-550c-438d-9f36-beb718bd3da8.png)
![snapshot-$j-13:11:30](https://user-images.githubusercontent.com/114431859/234080831-da2b500f-3bda-48a4-9278-98ae5ccf0e8b.png)

No particular file names are necessary. The delegation string is plain text, and the final delegation file is in hex, here's an example:
[Nostr_Key-A-delegation-4.txt](https://github.com/vitorpamplona/amethyst/files/11313391/Nostr_Key-A-delegation-4.txt)

**Comment by cobratbq** [comment]
I would be interested in something like this. For [nostrkey](<https://codeberg.org/walletkey/nostrkey>), it should be easy to produce the delegation token and offer it as QR-code. It would be awesome to have such a simple set-up to use a (hardware-supported) key.

**Comment by vitorpamplona** [comment]
Closing this since NIP-26 has been deprecated.
